File: 2023/day07/ruby/solution.rb
require_relative '../../../utils/ruby/Aoc'

input = Aoc.get_input(year: 2023, day: 7)
example = <<~EXAMPLE
32T3K 765
T55J5 684
KK677 28
KTJJT 220
QQQJA 483
EXAMPLE

example = example.split("\n").map(&:strip)

TYPES = [50000, 41000, 32000, 31100, 22100, 21110, 11111]

def parse(input)
    input.map do |line|
        hand, bid = line.split(' ').map(&:strip)
        [hand, bid.to_i]
    end
end

def type(hand)
    hand.chars.tally.values.sort.reverse.join.ljust(5, '0').to_i
end

def card_order_values(hand, cards_order)
    hand.chars.map do |card|
        cards_order.size - cards_order.index(card)
    end
end

def part1(input)
    parse(input)
    .map { |hand, bid| [type(hand), hand, bid] }
    .sort_by { |type, hand, bid| [type, card_order_values(hand, "AKQJT98765432")] }
    .each.with_index.sum { |(type, hand, bid), i| bid * (i + 1) }
end

# All possibilities to select n out of pool
# Each element can be selected multiple times
def selection(n, pool)
    return [[]] if n == 0

    pool.each.with_index.flat_map do |card, i|
        selection(n - 1, pool[i..-1]).map { |s| [card] + s }
    end
end

def type2(hand)
    hand_without_joker = hand.gsub('J', '')
    possible_joker_values = hand_without_joker.chars.uniq
    return TYPES.first if possible_joker_values.size <= 1

    selection(5 - hand_without_joker.size, possible_joker_values).map do |joker_values|
        type(hand_without_joker + joker_values.join)
    end.max
end

def part2(input)
    parse(input)
    .map { |hand, bid| [type2(hand), hand, bid] }
    .sort_by { |type, hand, bid| [type, card_order_values(hand, "AKQT98765432J")] }
    .each.with_index.sum { |(type, hand, bid), i| bid * (i + 1) }
end

puts "Part 1 (Example): #{part1(example)}"
puts "Part 1: #{part1(input)}"
puts "Part 2 (Example): #{part2(example)}"
puts "Part 2: #{part2(input)}"

File: 2023/day07/clojure/solution.clj
(ns solution)

(require '[clojure.string :as str])
(require '[babashka.deps :as deps])

(deps/add-deps '{:deps
                 {org.clojure/math.combinatorics
                  {:mvn/version "0.2.0"}}})

(require '[clojure.math.combinatorics :as combo])

(def input (slurp "/Users/dob/code/advent-of-code/2023/day07/input.txt"))

(def example
  "32T3K 765
   T55J5 684
   KK677 28
   KTJJT 220
   QQQJA 483")

(def cards "AKQT98765432J")
(def types (list :five_of_a_kind :four_of_a_kind :full_house :three_of_a_kind, :two_pairs, :one_pair, :high_card))

(defn position [item list]
  (loop [item item
         [head & tail] list
         idx 0]
    (cond
      (= item head) idx
      (empty? tail) nil
      :else (recur item tail (inc idx)))))

(defn type-of-hand [cards]
  (let [freq (->> cards frequencies vals sort)]
    (condp = freq
      '(5) :five_of_a_kind
      '(1 4) :four_of_a_kind
      '(2 3) :full_house
      '(1 1 3) :three_of_a_kind
      '(1 2 2) :two_pairs
      '(1 1 1 2) :one_pair
      :high_card)))

(defn cards-comparator
  ([] 0)
  ([[a & a_tail], [b & b_tail]]
   (cond
     (= a b) (cards-comparator a_tail b_tail)
     :else (- (str/index-of cards a) (str/index-of cards b)))))

(defn hands-comparator
  ([] 0)
  ([a, b]
  ;;  (let [a_type (type-of-hand-with-joker a),
  ;;        b_type (type-of-hand-with-joker b)]
   (let [a_type (type-of-hand a),
         b_type (type-of-hand b)]
     (cond
       (= a_type b_type) (cards-comparator a b)
       :else (- (position a_type types) (position b_type types))))))

(defn type-of-hand-with-joker
  "Highest type  "
  ([hand] (let [hand-without-joker (remove #{\J} hand),
                distinct-cards (distinct hand-without-joker),
                n (- 5 (count hand-without-joker)),
                possible-replacements (combo/selections distinct-cards n),
                possible-hands (map #(concat hand-without-joker %) possible-replacements)]
            (if (empty? distinct-cards)
              :five_of_a_kind
              (->> possible-hands
                   (sort hands-comparator)
                   first
                   type-of-hand)))))

(defn parse [input]
  (let [lines (str/split-lines input)]
    (->> lines
         (map str/trim)
         (map #(str/split % #" ")))))

(defn part1 [input]
  (->> input
       (parse)
       (sort-by first hands-comparator)
       reverse
       (map #(. Integer parseInt (second %)))
       (map-indexed (fn [idx bid] (* (inc idx) bid)))
       (reduce +)))

(defn part2 [input] (part1 input))


(println "Part 1")
(println (list "Example" (part1 example)))
(println (list "Input" (part1 input)))

(defn hands-comparator
  ([] 0)
  ([a, b]
   (let [a_type (type-of-hand-with-joker a),
         b_type (type-of-hand-with-joker b)]
  ;;  (let [a_type (type-of-hand a),
  ;;        b_type (type-of-hand b)]
     (cond
       (= a_type b_type) (cards-comparator a b)
       :else (- (position a_type types) (position b_type types))))))

(defn type-of-hand-with-joker
  "Highest type  "
  ([hand] (let [hand-without-joker (remove #{\J} hand),
                distinct-cards (distinct hand-without-joker),
                n (- 5 (count hand-without-joker)),
                possible-replacements (combo/selections distinct-cards n),
                possible-hands (map #(concat hand-without-joker %) possible-replacements)]
            (if (empty? distinct-cards)
              :five_of_a_kind
              (->> possible-hands
                   (sort hands-comparator)
                   first
                   type-of-hand)))))

(println "Part 2")
(println (list "Example" (part2 example)))
(println (list "Input" (part2 input)))